File: Services/CodexTaskRecoveryService.cs
using System.Diagnostics;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace CodexAccountBar.Services;

public sealed record RecoverableCodexTask(string Id, string Title, string WorkingDirectory);
public sealed record CodexTaskRecoveryResult(int Attempted, int Started, IReadOnlyList<string> Failures);

public sealed class CodexTaskRecoveryService(string codexHome, string codexExecutable)
{
    public async Task<IReadOnlyList<RecoverableCodexTask>> FindUsageLimitedTasksAsync(
        DateTimeOffset? since = null,
        int maximumCount = 5
    )
    {
        await using var server = await AppServerProbe.StartAsync(codexHome, codexExecutable);
        return await server.FindUsageLimitedTasksAsync(
            since ?? DateTimeOffset.UtcNow.AddMinutes(-30),
            maximumCount
        );
    }

    public Task<CodexTaskRecoveryResult> ResumeAsync(
        IReadOnlyList<RecoverableCodexTask> tasks
    )
    {
        var started = 0;
        var failures = new List<string>();
        const string prompt = "The previous account ran out of Codex quota. Continue this task from where it stopped. Preserve existing work, finish the original objective, and verify the result before stopping.";
        foreach (var task in tasks)
        {
            try
            {
                ProcessRunner.LaunchDetached(
                    codexExecutable,
                    ["exec", "resume", "--skip-git-repo-check", task.Id, prompt],
                    env: new Dictionary<string, string?> { ["CODEX_HOME"] = codexHome },
                    workingDirectory: Directory.Exists(task.WorkingDirectory)
                        ? task.WorkingDirectory
                        : null
                );
                started++;
            }
            catch (Exception exception)
            {
                failures.Add($"{task.Title}: {exception.Message}");
            }
        }
        return Task.FromResult(new CodexTaskRecoveryResult(tasks.Count, started, failures));
    }

    private sealed class AppServerProbe : IAsyncDisposable
    {
        private readonly Process _process;
        private readonly CancellationTokenSource _deadline = new(TimeSpan.FromSeconds(30));
        private int _nextRequestId = 1;

        private AppServerProbe(Process process) => _process = process;

        public static async Task<AppServerProbe> StartAsync(string codexHome, string executable)
        {
            var info = new ProcessStartInfo(executable)
            {
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                WindowStyle = ProcessWindowStyle.Hidden
            };
            info.ArgumentList.Add("app-server");
            info.Environment["CODEX_HOME"] = codexHome;
            var process = Process.Start(info)
                ?? throw new InvalidOperationException("Could not start Codex app-server.");
            _ = process.StandardError.ReadToEndAsync();
            var probe = new AppServerProbe(process);
            await probe.RequestAsync("initialize", new
            {
                clientInfo = new
                {
                    name = "codex_account_bar",
                    title = "Codex Account Bar",
                    version = "2.2.0"
                },
                capabilities = new { experimentalApi = true }
            });
            await probe.SendAsync(new { method = "initialized", @params = new { } });
            return probe;
        }

        public async Task<IReadOnlyList<RecoverableCodexTask>> FindUsageLimitedTasksAsync(
            DateTimeOffset since,
            int maximumCount
        )
        {
            var list = await RequestAsync("thread/list", new
            {
                limit = 30,
                sortKey = "updated_at",
                sortDirection = "desc",
                sourceKinds = new[] { "vscode", "cli", "appServer", "exec" }
            });
            var cutoff = since.ToUnixTimeSeconds();
            var rows = list["data"]?.AsArray() ?? [];
            var tasks = new List<RecoverableCodexTask>();
            foreach (var node in rows)
            {
                if (tasks.Count >= maximumCount || node is not JsonObject row) break;
                if ((row["updatedAt"]?.GetValue<long>() ?? 0) < cutoff) continue;
                if (row["parentThreadId"] is not null) continue;
                var threadId = row["id"]?.GetValue<string>();
                if (string.IsNullOrWhiteSpace(threadId)) continue;

                var turnsPage = await RequestAsync("thread/turns/list", new
                {
                    threadId,
                    limit = 1,
                    sortDirection = "desc",
                    itemsView = "summary"
                });
                var turns = turnsPage["data"]?.AsArray();
                if (turns is null || turns.Count == 0 || turns[^1] is not JsonObject lastTurn) continue;
                var errorInfo = lastTurn["error"]?["codexErrorInfo"]?.GetValue<string>();
                if (lastTurn["status"]?.GetValue<string>() != "failed"
                    || errorInfo != "usageLimitExceeded") continue;

                var title = FirstNonEmpty(
                    row["name"]?.GetValue<string>(),
                    row["preview"]?.GetValue<string>(),
                    threadId
                );
                if (title.Length > 80) title = title[..80];
                tasks.Add(new RecoverableCodexTask(
                    threadId,
                    title,
                    row["cwd"]?.GetValue<string>() ?? Environment.CurrentDirectory
                ));
            }
            return tasks;
        }

        private async Task<JsonObject> RequestAsync(string method, object parameters)
        {
            var id = _nextRequestId++;
            await SendAsync(new { method, id, @params = parameters });
            while (await _process.StandardOutput.ReadLineAsync(_deadline.Token) is { } line)
            {
                var message = JsonNode.Parse(line)?.AsObject();
                if (message?["id"]?.GetValue<int>() != id) continue;
                if (message["error"] is JsonObject error)
                {
                    throw new InvalidOperationException(
                        error["message"]?.GetValue<string>() ?? "Unknown Codex app-server error."
                    );
                }
                return message["result"]?.AsObject()
                    ?? throw new InvalidDataException($"Codex returned an invalid response for {method}.");
            }
            throw new EndOfStreamException("Codex app-server stopped unexpectedly.");
        }

        private async Task SendAsync(object message)
        {
            await _process.StandardInput.WriteLineAsync(JsonSerializer.Serialize(message));
            await _process.StandardInput.FlushAsync();
        }

        public async ValueTask DisposeAsync()
        {
            try { _process.StandardInput.Close(); } catch { }
            if (!_process.HasExited)
            {
                try { _process.Kill(true); } catch { }
                try { await _process.WaitForExitAsync(); } catch { }
            }
            _process.Dispose();
            _deadline.Dispose();
        }

        private static string FirstNonEmpty(params string?[] values) =>
            values.First(value => !string.IsNullOrWhiteSpace(value))!;
    }
}

File: macOS/Sources/CodexAccountBar/CodexTaskRecoveryService.swift
import Foundation

struct RecoverableCodexTask: Sendable, Hashable {
    let id: String
    let title: String
    let workingDirectory: String
}

struct CodexTaskRecoveryResult: Sendable {
    let attempted: Int
    let started: Int
    let failures: [String]
}

struct CodexTaskRecoveryService: Sendable {
    private let codexHome: URL
    private let codexExecutable: URL

    init(codexHome: URL, codexExecutable: URL) {
        self.codexHome = codexHome
        self.codexExecutable = codexExecutable
    }

    func findUsageLimitedTasks(
        since: Date = Date().addingTimeInterval(-30 * 60),
        maximumCount: Int = 5
    ) async throws -> [RecoverableCodexTask] {
        let codexHome = codexHome
        let executable = codexExecutable
        return try await Task.detached {
            try AppServerProbe(
                codexHome: codexHome,
                executable: executable
            ).findUsageLimitedTasks(since: since, maximumCount: maximumCount)
        }.value
    }

    func resume(_ tasks: [RecoverableCodexTask]) async -> CodexTaskRecoveryResult {
        guard !tasks.isEmpty else {
            return CodexTaskRecoveryResult(attempted: 0, started: 0, failures: [])
        }
        var started = 0
        var failures: [String] = []
        var environment = ProcessInfo.processInfo.environment
        environment["CODEX_HOME"] = codexHome.path
        let prompt = "The previous account ran out of Codex quota. Continue this task from where it stopped. Preserve existing work, finish the original objective, and verify the result before stopping."

        for task in tasks {
            do {
                let directory = URL(fileURLWithPath: task.workingDirectory, isDirectory: true)
                _ = try ProcessRunner.launchDetached(
                    codexExecutable,
                    arguments: [
                        "exec", "resume", "--skip-git-repo-check",
                        task.id, prompt
                    ],
                    environment: environment,
                    currentDirectoryURL: directory
                )
                started += 1
            } catch {
                failures.append("\(task.title): \(error.localizedDescription)")
            }
        }
        return CodexTaskRecoveryResult(
            attempted: tasks.count,
            started: started,
            failures: failures
        )
    }
}

private final class AppServerProbe {
    private let codexHome: URL
    private let executable: URL
    private var nextRequestID = 1

    init(codexHome: URL, executable: URL) {
        self.codexHome = codexHome
        self.executable = executable
    }

    func findUsageLimitedTasks(
        since: Date,
        maximumCount: Int
    ) throws -> [RecoverableCodexTask] {
        let process = Process()
        let input = Pipe()
        let output = Pipe()
        var environment = ProcessInfo.processInfo.environment
        environment["CODEX_HOME"] = codexHome.path
        process.executableURL = executable
        process.arguments = ["app-server"]
        process.environment = ProcessRunner.preparedEnvironment(environment)
        process.standardInput = input
        process.standardOutput = output
        process.standardError = FileHandle.nullDevice
        try process.run()
        let deadline = ProcessDeadline(process: process)
        deadline.schedule(after: 30)
        defer {
            deadline.cancel()
            try? input.fileHandleForWriting.close()
            if process.isRunning { process.terminate() }
        }

        let reader = JSONLineReader(handle: output.fileHandleForReading)
        _ = try request(
            method: "initialize",
            params: [
                "clientInfo": [
                    "name": "codex_account_bar",
                    "title": "Codex Account Bar",
                    "version": "2.2.0"
                ],
                "capabilities": ["experimentalApi": true]
            ],
            input: input.fileHandleForWriting,
            reader: reader
        )
        try send(
            ["method": "initialized", "params": [:]],
            to: input.fileHandleForWriting
        )
        let list = try request(
            method: "thread/list",
            params: [
                "limit": 30,
                "sortKey": "updated_at",
                "sortDirection": "desc",
                "sourceKinds": ["vscode", "cli", "appServer", "exec"]
            ],
            input: input.fileHandleForWriting,
            reader: reader
        )
        let cutoff = Int(since.timeIntervalSince1970)
        let rows = (list["data"] as? [[String: Any]] ?? []).filter { row in
            let updatedAt = (row["updatedAt"] as? NSNumber)?.intValue ?? 0
            let isRoot = row["parentThreadId"] == nil || row["parentThreadId"] is NSNull
            return updatedAt >= cutoff && isRoot
        }

        var tasks: [RecoverableCodexTask] = []
        for row in rows where tasks.count < maximumCount {
            guard let threadID = row["id"] as? String else { continue }
            let turnsPage = try request(
                method: "thread/turns/list",
                params: [
                    "threadId": threadID,
                    "limit": 1,
                    "sortDirection": "desc",
                    "itemsView": "summary"
                ],
                input: input.fileHandleForWriting,
                reader: reader
            )
            guard let turns = turnsPage["data"] as? [[String: Any]],
                  let lastTurn = turns.last,
                  lastTurn["status"] as? String == "failed",
                  let error = lastTurn["error"] as? [String: Any],
                  error["codexErrorInfo"] as? String == "usageLimitExceeded"
            else { continue }

            let name = (row["name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            let preview = (row["preview"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            let title = [name, preview].compactMap { $0 }.first(where: { !$0.isEmpty }) ?? threadID
            let workingDirectory = row["cwd"] as? String ?? codexHome.deletingLastPathComponent().path
            tasks.append(RecoverableCodexTask(
                id: threadID,
                title: String(title.prefix(80)),
                workingDirectory: workingDirectory
            ))
        }
        return tasks
    }

    private func request(
        method: String,
        params: [String: Any],
        input: FileHandle,
        reader: JSONLineReader
    ) throws -> [String: Any] {
        let requestID = nextRequestID
        nextRequestID += 1
        try send(
            ["method": method, "id": requestID, "params": params],
            to: input
        )
        while let message = try reader.nextObject() {
            guard (message["id"] as? NSNumber)?.intValue == requestID else { continue }
            if let error = message["error"] as? [String: Any] {
                throw CodexTaskRecoveryError.appServer(
                    error["message"] as? String ?? "Unknown app-server error"
                )
            }
            guard let result = message["result"] as? [String: Any] else {
                throw CodexTaskRecoveryError.invalidResponse(method)
            }
            return result
        }
        throw CodexTaskRecoveryError.appServerStopped
    }

    private func send(_ object: [String: Any], to handle: FileHandle) throws {
        var data = try JSONSerialization.data(withJSONObject: object)
        data.append(0x0A)
        try handle.write(contentsOf: data)
    }
}

private final class ProcessDeadline: @unchecked Sendable {
    private let lock = NSLock()
    private let process: Process
    private var cancelled = false

    init(process: Process) {
        self.process = process
    }

    func schedule(after seconds: TimeInterval) {
        DispatchQueue.global().asyncAfter(deadline: .now() + seconds) { [self] in
            lock.lock()
            let shouldTerminate = !cancelled
            lock.unlock()
            if shouldTerminate, process.isRunning { process.terminate() }
        }
    }

    func cancel() {
        lock.lock()
        cancelled = true
        lock.unlock()
    }
}

private final class JSONLineReader {
    private let handle: FileHandle
    private var buffer = Data()

    init(handle: FileHandle) {
        self.handle = handle
    }

    func nextObject() throws -> [String: Any]? {
        while true {
            if let newline = buffer.firstIndex(of: 0x0A) {
                let line = buffer[..<newline]
                buffer.removeSubrange(...newline)
                guard !line.isEmpty else { continue }
                return try JSONSerialization.jsonObject(with: Data(line)) as? [String: Any]
            }
            let chunk = handle.availableData
            guard !chunk.isEmpty else {
                return nil
            }
            buffer.append(chunk)
        }
    }
}

enum CodexTaskRecoveryError: LocalizedError {
    case appServer(String)
    case appServerStopped
    case invalidResponse(String)

    var errorDescription: String? {
        switch self {
        case .appServer(let message): "Codex task scan failed: \(message)"
        case .appServerStopped: "Codex task scan stopped unexpectedly."
        case .invalidResponse(let method): "Codex returned an invalid response for \(method)."
        }
    }
}
